File: crates/evobase-protocol/src/envelope.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct ApiResponse<T: Serialize> {
    pub data: T,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<ResponseMeta>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ResponseMeta {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<i64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ErrorEnvelope {
    pub error: ErrorDetail,
}

#[derive(Debug, Clone, Serialize)]
pub struct ErrorDetail {
    pub code: &'static str,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub field: Option<String>,
}

impl<T: Serialize> ApiResponse<T> {
    pub fn new(data: T) -> Self {
        Self { data, meta: None }
    }

    pub fn with_meta(data: T, meta: ResponseMeta) -> Self {
        Self {
            data,
            meta: Some(meta),
        }
    }
}

File: apps/evobase_workbench/lib/client/models/envelope.dart
import 'package:json_annotation/json_annotation.dart';

part 'envelope.g.dart';

/// Mirror of evobase-protocol ApiResponse
@JsonSerializable(genericArgumentFactories: true)
class ApiResponse<T> {
  final T data;
  final ResponseMeta? meta;

  ApiResponse({required this.data, this.meta});

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) => _$ApiResponseFromJson(json, fromJsonT);

  Map<String, dynamic> toJson(Object Function(T value) toJsonT) =>
      _$ApiResponseToJson(this, toJsonT);
}

@JsonSerializable()
class ResponseMeta {
  final int? total;
  final int? limit;
  final int? offset;

  ResponseMeta({this.total, this.limit, this.offset});

  factory ResponseMeta.fromJson(Map<String, dynamic> json) =>
      _$ResponseMetaFromJson(json);

  Map<String, dynamic> toJson() => _$ResponseMetaToJson(this);
}

/// Mirror of evobase-protocol ErrorEnvelope
@JsonSerializable()
class ErrorEnvelope {
  final ErrorDetail error;

  ErrorEnvelope({required this.error});

  factory ErrorEnvelope.fromJson(Map<String, dynamic> json) =>
      _$ErrorEnvelopeFromJson(json);

  Map<String, dynamic> toJson() => _$ErrorEnvelopeToJson(this);
}

@JsonSerializable()
class ErrorDetail {
  final String code;
  final String message;
  final String? field;

  ErrorDetail({required this.code, required this.message, this.field});

  factory ErrorDetail.fromJson(Map<String, dynamic> json) =>
      _$ErrorDetailFromJson(json);

  Map<String, dynamic> toJson() => _$ErrorDetailToJson(this);
}
